### ISSUE TITLE
Classpath is incomplete from 0.6.4

### ISSUE THREAD

**Issue by ppettit** [atom/ide-java#37]
With 0.6.0 I was using ide-java with no issues (other than the auto complete bug #10 and taking 4 minutes for JDT to start up) with a big maven [project](https://bitbucket.org/opencast-community/opencast).

0.6.1-3 didn't work due to #30 
0.6.4-5 give me the "Classpath is incomplete" dialog

I guess the change from Eclipse JDT 0.3 to 0.5 has broken something?

There are no errors in the dev console. Tried with Atom 1.21 and 1.22.

**Comment by n00dl3** [comment]
I have same problem tough I open a folder containing a `pom.xml`.
![2017-10-11-110053_956x505_scrot](https://user-images.githubusercontent.com/3089875/31431349-ba332758-ae73-11e7-9cdc-85919bbf37a3.png)

**Comment by ppettit** [comment]
I read the page, but that does not help as the project has a pom.xml file in the root just as @n00dl3 commented above.

Also, as mentioned, it works fine with ide-java 0.6.0. just downgrading the package with apm and restarting atom makes it work again.

Anything else I can try?

**Comment by tylerFowler** [comment]
Had the same issue, is there any way to enable verbose logging? As mentioned nothing is logged to the atom console so I'm not even sure where to start troubleshooting short of cleaning & reinstalling.

Same though with downgrading to v0.6.0, that seems to work.

**Comment by remigastaldi** [comment]
Same issue but downgrading to v0.6.0 give me "Uncaught Error: Header must provide a Content-Length property."

**Comment by ajoberstar** [comment]
I've been researching this one as well because I haven't gotten a single Gradle project to work. From what I'm finding, you should always get incomplete classpath for all Maven and Gradle projects.

The offending commit in ide-java being [the upgrade of the Eclipse JDT](https://github.com/atom/ide-java/commit/8dff75163557a95caaf4d40835948f467ee4e5b1). Though technically, this is really an issue of mismatching LSP usage between atom-languageclient and the Eclipse JDT LS.

## Issue

- The language server protocol defines both a `rootUri` and `rootPath` that can be passed in during `initialize`. As of LSP 3.0 `rootPath` is deprecated in favor of `rootUri`.
- The Eclipse JDT Server uses `rootUri` (and currently ignores `rootPath`) to determine your project's root, which is where it scans for the presence of `build.gradle` or `pom.xml` to support Gradle and Maven projects. 
- atom-languageclient has an InitializeParams type which only supports a `rootPath`.

As a result, JDT never sees the root of your project and just sets up a barebones Java project.

## Proposed Solutions

Both of these changes should probably happen, but either one would fix ide-java (once ide-java upgrades to the new versions of these dependencies).

- atom-languageclient should pass both `rootPath` and `rootUri` (as [vscode-languageclient does](https://github.com/Microsoft/vscode-languageserver-node/blob/42a04f600f2d023ef2ba81002f4235e22e5a54fe/client/src/client.ts#L2499))
- Eclipse JDT should check both `rootPath` and `rootUri` (only using `rootPath`, if rootUri is null)

I'll open issues for both of these and hopefully one or both will be resolved.

## Research

### JDT LS Log

from `%TEMP%\atom-java-<encoded project path>\.metadata\.log`

```
!ENTRY org.eclipse.jdt.ls.core 1 0 2017-10-22 12:05:40.176
!MESSAGE >> initialize

!ENTRY org.eclipse.jdt.ls.core 1 0 2017-10-22 12:05:40.178
!MESSAGE Initializing Java Language Server 0.5.0.201709251422

!ENTRY org.eclipse.jdt.ls.core 1 0 2017-10-22 12:05:40.178
!MESSAGE No root uri was defined. Falling back on C:\Users\andre\AppData\Local\Temp\atom-java-C%3A%5Cusers%5Candre%5Cprojects%5Creckon%5C
```

Key message being **no root uri was defined**

### atom-languageclient

Enabling `core.debugLSP` in Atom config gives messages showing the communication between Atom and the language server as console logs in developer tools.

```
Java (Eclipse JDT) rps.sendRequest initialize sending
Object
  ...
  rootPath: "C:\users\andre\projects\reckon\"
  ...
``` 

atom-languageclient sends this intialization message itself by default (ide-java has not overriden this) [in getInitializeParams](https://github.com/atom/atom-languageclient/blob/v0.6.4/lib/auto-languageclient.js#L78). The [InitializeParams](https://github.com/atom/atom-languageclient/blob/e37fc73eeb956adb8c4e1b911e7d21f29f55ea76/lib/languageclient.js#L358) type, also part of atom-languageclient, specifies only the `rootPath` as an option.

### Eclipse JDT

In the 0.5.0 JDT [currently being used](https://github.com/atom/ide-java/blob/77851199271e94d83eadfff01fc414972555389a/lib/main.js#L8), that _no root uri_ message occurs [here](https://github.com/eclipse/eclipse.jdt.ls/blob/v0.5.0/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/InitHandler.java#L83). Above that line you can see it checking `InitializeParams#getRootUri()` to determine the root URI.

Notably, Atom is sending `rootPath` and JDT is checking `rootUri`. [This JDT commit](https://github.com/eclipse/eclipse.jdt.ls/commit/277ee177090843c0b2560e779d3bff3af9dd9ce5#diff-77a4e07ecb50e4529272a8d107e6c143) changed logic in 0.1.0 that used to check both `rootPath` and `rootUri`. This behavior is still true in master of JDT.

Between 0.6.0 and 0.6.1 of ide-java, The JDT dependency went from an old not final 0.1.0 (from April) to 0.3.0 (and has since moved beyond that version).

### LSP4J

`IntializeParams` is part of LSP4J.

Currently LSP4J 0.3.0 is being used by JDT. [InitializeParams shows](https://github.com/eclipse/lsp4j/blob/3c309e0fe63056bb8146c104774c0ebb2112cb58/org.eclipse.lsp4j/src/main/xtend-gen/org/eclipse/lsp4j/InitializeParams.java#L77) both a `rootUri` and a `rootPath` (`rootPath being deprecated). That deprecation was introduced in [a commit targeting compatilbility with LSP 3.0](https://github.com/eclipse/lsp4j/commit/3c309e0fe63056bb8146c104774c0ebb2112cb58).

### Language Server Protocol

[LSP 3.0 indicates](https://github.com/Microsoft/language-server-protocol/blob/3.0.0/protocol.md#initialize) that rootDir has been deprecated in favor of rootUri and that if both are defined, `rootUri` wins.

**Comment by damieng** [comment]
Thanks for researching this!

I was trying to hold off LSP 3.0 support and there wasn't any indication the new language server was breaking 2.0 support.

I can add support for passing both as a subset of 3.0 support for now.

**Comment by damieng** [comment]
Fixed in 0.6.6!  Sorry for the delay.
